junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/DefaultLauncher.java

```diff
@@ -72,6 +79,32 @@
 		return testEngines;
 	}
 
+	// https://github.com/junit-team/junit5/issues/1557
+	private static boolean validateReservedIds(TestEngine testEngine) {
+		String engineId = testEngine.getId();
+		if (!engineId.startsWith("junit-")) {
+			return true;
+		}
+		if (engineId.equals("junit-jupiter")) {
+			validateWellKnownTypeName(testEngine, "org.junit.jupiter.engine.JupiterTestEngine");
+			return true;
+		}
+		if (engineId.equals("junit-vintage")) {
+			validateWellKnownTypeName(testEngine, "org.junit.vintage.engine.VintageTestEngine");
+			return true;
+		}
+		return false;
+	}
+
+	private static void validateWellKnownTypeName(TestEngine testEngine, String expectedTypeName) {
+		String actualTypeName = testEngine.getClass().getTypeName();
+		if (actualTypeName.equals(expectedTypeName)) {
+			return;
+		}
+		throw new JUnitException(
+			String.format("Impostor detected! Expected %s but got: %s", expectedTypeName, actualTypeName));
```

Comment:
Hos about something more like: _Third-party TestEngine [FQCN] is forbidden to use the reserved 'junit-vintage' TestEngine ID._